### ISSUE TITLE
cppcoreguidelines-narrowing-conversions output the opposite warning for float conversion

### ISSUE THREAD

**Issue by 7fa3d1d8-8ed8-4218-a382-cf1f0c3e5f27** [llvm/llvm-project#47554]
|  |  |
| --- | --- |
| Bugzilla Link | [48210](https://llvm.org/bz48210) |
| Version | unspecified |
| OS | Linux |
| CC | @legrosbuffle,@EugeneZelenko |

## Extended Description 
cppcoreguidelines-narrowing-conversions is reporting narrowing warnings for converting smaller integer types into float that cannot cause narrowing yet not reporting warnings for bigger types that can cause narrowing.

Example code:
------
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <limits>

using IntType = int8_t;

float foo(IntType val) { return val + 0.1F; }
float bar(IntType val) { return val * 2.0f; }

int main() {
  const IntType val = std::numeric_limits<IntType>::max();

  std::cout.precision(20);
  std::cout << static_cast<intmax_t>(val) << '\n';
  std::cout << foo(val) << '\n';
  std::cout << bar(val) << '\n';
}
------

clang-tidy output:
------
$ clang-tidy -checks="-*, cppcoreguidelines-narrowing-conversions" test.cpp --
3 warnings generated.
/workspace/test.cpp:8:33: warning: narrowing conversion from 'int' to 'float' [cppcoreguidelines-narrowing-conversions]
float foo(IntType val) { return val + 0.1F; }
                                ^
/workspace/test.cpp:9:33: warning: narrowing conversion from 'int' to 'float' [cppcoreguidelines-narrowing-conversions]
float bar(IntType val) { return val * 2.0f; }
------

The same warnings are reported for (u)int8_t and (u)int16_t. No warnings are reported for (u)int32_t and (u)int64_t. The check output seems backwards, reporting a warning when there is no narrowing but no warning when there is narrowing.

The warning message is also confusing, reporting the type after integer promotion. It seems it's not keeping track of the original type before integer promotion?

Observed on clang-tidy installed from repository on Ubuntu 20.04

$ clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 10.0.0
  
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: skylake

**Comment by 7fa3d1d8-8ed8-4218-a382-cf1f0c3e5f27** [comment]
Interestingly clang itself gives the correct warning with -Weverything:

uint32_t:

```console
$ clang++ test.cpp -std=c++17 -Weverything -Wno-c++98-compat
test.cpp:8:40: warning: implicit conversion from 'IntType' (aka 'unsigned int') to 'float' may lose precision [-Wimplicit-int-float-conversion]
static float foo(IntType val) { return val + 0.1F; }
                                       ^~~ ~
test.cpp:9:40: warning: implicit conversion from 'IntType' (aka 'unsigned int') to 'float' may lose precision [-Wimplicit-int-float-conversion]
static float bar(IntType val) { return val * 2.0f; }
                                       ^~~ ~
2 warnings generated.
```

uint16_t:

```console
$ clang++ test.cpp -std=c++17 -Weverything -Wno-c++98-compat
```

**Comment by vbvictor** [comment]
Fixed on trunk https://godbolt.org/z/WEoo8T165
